packages/contracts-bedrock/src/safe/LivenessModule2.sol

```diff
@@ -0,0 +1,226 @@
+            revert LivenessModule2_ModuleNotConfigured();
+        }
+
+        // Check that this module is NOT enabled on the calling Safe
+        // This prevents clearing configuration while module is still enabled
+        Safe safe = Safe(payable(msg.sender));
+        if (safe.isModuleEnabled(address(this))) {
+            revert LivenessModule2_ModuleStillEnabled();
+        }
+
+        // Erase the configuration data for this safe
+        delete safeConfigs[msg.sender];
+        // Also clear any active challenge
+        delete challengeStartTime[msg.sender];
+        emit ModuleDisabled(msg.sender);
+    }
+
+    /// @notice Returns challenge_start_time + liveness_response_period if there is a challenge for the given safe, or
+    /// 0 if not
+    /// @dev MUST never revert
+    /// @param _safe The Safe address to query
+    /// @return The challenge end timestamp, or 0 if no challenge
+    function getChallengePeriodEnd(address _safe) external view returns (uint256) {
+        uint256 startTime = challengeStartTime[_safe];
+        if (startTime == 0) {
+            return 0;
+        }
+        ModuleConfig storage config = safeConfigs[_safe];
+        return startTime + config.livenessResponsePeriod;
+    }
+
+    /// @notice Challenges an enabled safe
+    /// @dev MUST only be executable by fallback owner of the challenged safe
+    /// @dev MUST revert if there is a challenge for the safe
+    /// @dev MUST set challenge_start_time to the current block time
+    /// @dev MUST emit the ChallengeStarted event
+    /// @param _safe The Safe to challenge
+    function challenge(address _safe) external {
+        ModuleConfig storage config = safeConfigs[_safe];
+
+        if (config.fallbackOwner == address(0)) {
+            revert LivenessModule2_ModuleNotConfigured();
+        }
+
+        // Check that the module is still enabled on the target Safe
+        Safe safe = Safe(payable(_safe));
+        if (!safe.isModuleEnabled(address(this))) {
+            revert LivenessModule2_ModuleNotEnabled();
+        }
+
+        if (msg.sender != config.fallbackOwner) {
+            revert LivenessModule2_UnauthorizedCaller();
+        }
+
+        if (challengeStartTime[_safe] != 0) {
+            revert LivenessModule2_ChallengeAlreadyExists();
+        }
+
+        challengeStartTime[_safe] = block.timestamp;
+        emit ChallengeStarted(_safe, block.timestamp);
+    }
+
+    /// @notice Responds to a challenge for an enabled safe, canceling it
+    /// @dev MUST only be executable by an enabled safe
+    /// @dev MUST revert if there isn't a challenge for the calling safe
+    /// @dev MUST revert if there is a challenge for the calling safe but the response period has expired
+    /// @dev MUST emit the ChallengeCancelled event
+    function respond() external {
+        // Check that this module is enabled on the calling Safe
+        Safe safe = Safe(payable(msg.sender));
+        if (!safe.isModuleEnabled(address(this))) {
```

Comment:
We do this check + the fallback owner check repeatedly, we can probably use a single `isModuleActive(safe)` function instead.